design/complex-workloads.md

```diff
@@ -0,0 +1,988 @@
+* Each `Workload` is a unit of scheduling; it may not be spread across multiple
+  `KubernetesClusters`.
+* The proposed design lay a foundation for supporting workloads that are not
+  containerised.
+
+The following are out of scope for the Workload resource:
+* Deploying a single workload to multiple clusters simultaneously.
+* Configuration and/or templating. Each `Workload` will be a 'static' resource;
+  the task of generating or altering `Workloads` given a set of inputs will
+  be that of a higher level construct.
+* Package and dependency management. `Workloads` will not model dependencies on
+  or relationships to other `Workloads`. Any resource types upon which a
+  `Workload` depends are presumed to have been defined via CRD before
+  instantiating the `Workload`.
+
+## Design
+
+### Custom Resource Definitions
+This document proposes the `Workload` kind within the
+`compute.crossplane.io/v1alpha1` API group be renamed to the
+`KubernetesApplication` kind in the `workload.crossplane.io/v1alpha` group.
+`.spec.targetDeployment`, `.spec.targetService`, `.spec.targetNamespace`, and
+all fields of `.status` that relate to a `Deployment` or `Service` be removed.
+These concrete types will be replaced by a new 'envelope' resource named
+`KubernetesApplicationComponent`. A relationship will be established between
+one `KubernetesApplication` and many `KubernetesApplicationComponents` via a
+[label selector][2].
+
+#### Schema
+Each `KubernetesApplicationComponent` represents a single Kubernetes resource to
+be deployed to a `KubernetesCluster`. The `KubernetesApplicationComponent`
+encapsulates the resource, including type and object metadata, in its
+`.spec.template` field. If the templated resource kind exposes a `.status` field
+when deployed, said field will be copied verbatim to the
+`KubernetesApplicationComponent`'s `.status.remote` field.
+`KubernetesApplicationComponents` will also specify a list of `Secrets` presumed
+to be the automatically created resource connection secrets for Crossplane
+managed resources upon which its templated Kubernetes resource depends. These
+`Secrets` will be propagated into the same namespace as the templated resource.
+
+Crossplane will model the template using the
+[`*unstructured.Unstructured`][3] type internally. Unstructured types must
+include Kubernetes type and object metadata but are otherwise opaque. Status
+will be completely opaque - i.e. a [`json.RawMessage`][4] - to the controller
+code. The controller will copy the remote resource's `.status` field into the 
+`KubernetesApplicationComponent`'s `.status`.remote field. `.status.remote`
+will be absent from `KubernetesApplicationComponents` that template resource
+kinds that do not expose a `.status` field.
+
+An example complex workload:
+```yaml
+---
+apiVersion: workload.crossplane.io/v1alpha1
+kind: KubernetesApplication
+metadata:
+  name: demo
+spec:
+  clusterSelector:
+    matchLabels:
+      provider: gcp
+  workSelector:
```

Comment:
TODO(negz): Rename this to `componentSelector`. This is an artifact of a stranger time in which I was calling a `KubernetesApplicationComponent` a `Work`.

Reply:
This selector no longer exists under the latest design iteration, so I've removed it entirely.